Add generic cassandra code

arcusfelis opened the pull request:
This PR addresses #719 

Proposed changes include:

Generic code is used by all cassandra backends
At this point there is no cassandra modules yet
New configuration option cassandra_servers
Documentation (Basic-configuration and advanced-configuration/Cassandra)
Preffered way of calling Cassandra is prepared queries

New modules:
mongoose_cassandra - start/stop cassandra connections
mongoose_cassandra_sup - pooling and supervision
mongoose_cassandra_worker - maintains seestar connection, keeps prepared
queries, exports query functions

michalwski commented:
Can we have some basic test cases at this point checking if the connection to cassandra was established?

arcusfelis commented:
Dializer will fail because I've updated seestar which returns different tuple from start_link
https://travis-ci.org/esl/MongooseIM/jobs/119318336

So, I removed mod_mam_con_ca_arch and mod_mam_muc_ca_arch modules.

michalwski commented on apps/ejabberd/src/mongoose_cassandra_sup.erl:
```diff
@@ -0,0 +1,98 @@
+    delete_worker_pool(PoolName).
+
+start_link(PoolName, Config) ->
+    supervisor2:start_link({local, ?MODULE}, ?MODULE, [PoolName, Config]).
```
Why do we need supervisor2 here?

arcusfelis replied:
No idea, fixed. Remainings from the older version.

michalwski commented on apps/ejabberd/src/mongoose_cassandra_sup.erl:
```diff
@@ -0,0 +1,98 @@
+%%====================================================================
+
+create_worker_pool(PoolName) ->
+    pg2:create(group_name(PoolName)).
```
Why exactly do we need a distributed process group?

arcusfelis replied:
Pooling. cuesport can be used instead.

michalwski replied:
I'd rather see `poolboy` here (or sth else) to fail fast if there is no spare workers.

arcusfelis replied:
cuesport doesn't provide any way to get pids of all workers in the pool (without hacks).

michalwski replied:
But why do you need all pids?

arcusfelis replied:
For this:

```erlang
select_worker(PoolName, UserJID) ->                                                  
    case pg2:get_local_members(group_name(PoolName)) of                              
        [] ->                                                                        
            error({no_worker, PoolName});                                            
        Workers ->                                                                   
            N = erlang:phash2(UserJID, length(Workers)) + 1,                         
            lists:nth(N, Workers)                                                    
    end.  
```
Also, pg2 allows to register a process when connection is ready, not when the process starts.

michalwski replied:
Ok, I see. Why exactly do you want to get the same worker for the same user? For subsequent queries?

arcusfelis replied:
poolboy is useless because cassandra queries are async. So, all workers are spare.

michalwski commented on apps/ejabberd/src/mongoose_cassandra_worker.erl:
```diff
@@ -0,0 +1,480 @@
+
+%% @doc Do CQL query
+%%
+%% UserJID is used for rate-limiting, statistics and debugging
```
If UseJID is used for debugging then maybe it's a good idea to produce a log message at lest in case of error with it? Currently I see it ignored everywhere.

michalwski replied:
You mention statistics also, but I can't see any cassandra related metrics. Are you going to add them in next PR?

arcusfelis replied:
Meaning of this JID is a partitioning key. Not a JID of user who makes the query. It can be used for some kind of locking, for example.
Currently it's only used in select_worker code. I.e. it's hashed to select worker.
I can add error reporting but backpressure and statistics is not that easy and require extending of the gen-server.
It also can be used to show warnings like this one:
```
issue=large_result_set, user=alice, number_of_elements=9000, query="SELECT ..."
```
The same information is inside query, but should be parsed. Here we can just use it.

michalwski commented on apps/ejabberd/src/mongoose_cassandra_worker.erl:
```diff
@@ -0,0 +1,480 @@
+            get_prepared_query_first_time(Conn, Module, QueryName, State)
+    end.
+
+get_prepared_query_first_time(Conn, Module, QueryName, State=#state{prepared_queries=PreparedQueries}) ->
```
This line looks a bit too long, could you split it into 2 lines?

michalwski commented on doc/advanced-configuration/Cassandra.md:
```diff
@@ -0,0 +1,70 @@
+## Advanced Cassandra configuration
```
You need to add this file to `mkdocs.yml` in order to have it included on readthedocs.org.

Also please try following command localy from repo's root 
```bash
mkdocs serve
```

to verify if everything's rendered correctly.

michalwski commented on test/ejabberd_tests/tests/cassandra_SUITE.erl:
```diff
@@ -0,0 +1,66 @@
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%==============================================================================
+-module(cassandra_SUITE).
```
How about moving this test suite to `small` tests? There is no XMPP/HTTP level tests here and that's why I think such a test should be placed in `apps/ejabberd/test` and not here.

In other words, IMO, such internal test cases not touching the end user interface of MongooseIM should be placed in `small` tests.

arcusfelis replied:
right

arcusfelis commented:
Closed, continue in #913
